rasterio/_err.pyx

```diff
@@ -39,17 +41,59 @@ cdef extern from "cpl_error.h":
     int CPLGetLastErrorType()
     void CPLErrorReset()
 
+
+class CPLError(Exception):
+    """Base CPL error class"""
+    pass
+
+
+class CPLE_AppDefined(CPLError):
+    pass
+
+
+class CPLE_OutOfMemory(CPLError):
+    pass
+
+
+class CPLE_FileIO(CPLError):
+    pass
+
+
+class CPLE_OpenFailed(CPLError):
+    pass
+
+
+class CPLE_IllegalArg(CPLError):
+    pass
+
+
+class CPLE_NotSupported(CPLError):
+    pass
+
+
+class CPLE_AssertionFailed(CPLError):
+    pass
+
+
+class CPLE_NoWriteAccess(CPLError):
+    pass
+
+
+class CPLE_UserInterrupt(CPLError):
+    pass
+
+
 # Map GDAL error numbers to Python exceptions.
 exception_map = {
-    1: RuntimeError,        # CPLE_AppDefined
-    2: MemoryError,         # CPLE_OutOfMemory
-    3: IOError,             # CPLE_FileIO
-    4: IOError,             # CPLE_OpenFailed
-    5: TypeError,           # CPLE_IllegalArg
-    6: ValueError,          # CPLE_NotSupported
-    7: AssertionError,      # CPLE_AssertionFailed
-    8: IOError,             # CPLE_NoWriteAccess
-    9: KeyboardInterrupt,   # CPLE_UserInterrupt
+    1: CPLE_AppDefined,        # 
+    2: CPLE_OutOfMemory,         # 
+    3: CPLE_FileIO,             # 
+    4: CPLE_OpenFailed,             # 
+    5: CPLE_IllegalArg,           # 
+    6: CPLE_NotSupported,          # 
+    7: CPLE_AssertionFailed,      # 
+    8: CPLE_NoWriteAccess,             # 
+    9: CPLE_UserInterrupt,   # 
```

Comment:
Should we go ahead and add the new vsi error messages here as well?

https://github.com/mapbox/rasterio/pull/591/files#diff-e13898fc6e575b9f0ed53e5d98208410R81

This would effectively close #591 as well